iceoryx_posh/include/iceoryx_posh/internal/popo/base_client.hpp

```diff
@@ -0,0 +1,147 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_BASE_CLIENT_HPP
+#define IOX_POSH_POPO_BASE_CLIENT_HPP
+
+#include "iceoryx_hoofs/cxx/expected.hpp"
+#include "iceoryx_posh/capro/service_description.hpp"
+#include "iceoryx_posh/internal/popo/ports/client_port_user.hpp"
+#include "iceoryx_posh/popo/client_options.hpp"
+#include "iceoryx_posh/popo/trigger_handle.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+namespace iox
+{
+namespace popo
+{
+using uid_t = UniquePortId;
+
+/// @brief The BaseClient class contains the common implementation for the different clients
+/// @param[in] Port type of the underlying port, required for testing
+template <typename Port = ClientPortUser>
+class BaseClient
+{
+  public:
+    virtual ~BaseClient() noexcept;
+
+    BaseClient(const BaseClient& other) = delete;
+    BaseClient& operator=(const BaseClient&) = delete;
+    BaseClient(BaseClient&& rhs) = delete;
+    BaseClient& operator=(BaseClient&& rhs) = delete;
+
+    ///
+    /// @brief Get the unique ID of the client.
+    /// @return The client's unique ID.
+    ///
+    uid_t getUid() const noexcept;
+
+    ///
+    /// @brief Get the service description of the client.
+    /// @return The service description.
+    ///
+    capro::ServiceDescription getServiceDescription() const noexcept;
+
+    ///
+    /// @brief Initiate connection to server.
+    /// @return
+    ///
+    void connect() noexcept;
+
+    ///
+    /// @brief Get current connection state.
+    /// @return The current connection state.
+    ///
+    ConnectionState getConnectionState() const noexcept;
+
+    ///
+    /// @brief Disconnects if currently connected, otherwise do nothing.
+    ///
+    void disconnect() noexcept;
+
+    ///
+    /// @brief Check if response are available.
+    /// @return True if responses are available.
+    ///
+    bool hasResponses() const noexcept;
+
+    ///
+    /// @brief Check if response has been missed since the last call of this method.
+    /// @return True if response has been missed.
+    /// @details Response may be missed due to overflowing receive queue.
+    ///
+    bool hasMissedResponses() noexcept;
+
+    /// @brief Releases any unread queued response.
+    void releaseQueuedResponses() noexcept;
+
+    friend class NotificationAttorney;
+
+  protected:
+    using SelfType = BaseClient<Port>;
+    using PortType = Port;
+
+    BaseClient(const capro::ServiceDescription& service, const ClientOptions& clientOptions) noexcept;
+
+    /// @brief Only usable by the WaitSet, not for public use. Invalidates the internal triggerHandle.
```

Comment:
Not important but: WaitSet and Listener. Goes also for the remaining comments.
The states on the other hand are only used by the waitset.